source/jormungandr/jormungandr/scenarios/ridesharing/ridesharing_service_manager.py

```diff
@@ -0,0 +1,342 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+# Stay tuned using
+# twitter @navitia
+# channel `#navitia` on riot https://riot.im/app/#/room/#navitia:matrix.org
+# https://groups.google.com/d/forum/navitia
+# www.navitia.io
+
+from jormungandr.scenarios.ridesharing.ridesharing_service import AbstractRidesharingService
+import logging, datetime
+from typing import Dict
+from jormungandr import utils
+import six
+from jormungandr import new_relic
+from jormungandr.scenarios import journey_filter
+from jormungandr.scenarios.ridesharing.ridesharing_journey import Gender
+from jormungandr.utils import get_pt_object_coord, generate_id
+from jormungandr.street_network.utils import crowfly_distance_between
+from navitiacommon import response_pb2
+from jormungandr.utils import PeriodExtremity
+from jormungandr.scenarios.journey_filter import to_be_deleted
+from importlib import import_module
+
+
+class RidesharingServiceManager(object):
+    def __init__(
+        self, instance, ridesharing_services_configuration, rs_services_getter=None, update_interval=60
+    ):
+        self.logger = logging.getLogger(__name__)
+        self.ridesharing_services_configuration = ridesharing_services_configuration
+        self._ridesharing_services_legacy = []
+        self._rs_services_getter = rs_services_getter
+        self._ridesharing_services = {}  # type: Dict[str, AbstractRidesharingService]
+        self._ridesharing_services_last_update = {}  # type: Dict[str, datetime]
+        self._last_update = datetime.datetime(1970, 1, 1)
+        self._update_interval = update_interval
+        self.instance = instance
+
+    def init_ridesharing_services(self):
+        # Init legacy ridesharing from config file
+        for config in self.ridesharing_services_configuration:
+            # Set default arguments
+            if 'args' not in config:
+                config['args'] = {}
+            if 'service_url' not in config['args']:
+                config['args'].update({'service_url': None})
+            try:
+                service = utils.create_object(config)
+            except KeyError as e:
+                raise KeyError(
+                    'impossible to build a ridesharing service for {}, '
+                    'missing mandatory field in configuration: {}'.format(self.instance.name, e.message)
+                )
+            self.logger.info(
+                '** Ridesharing: {} used for instance: {} **'.format(type(service).__name__, self.instance.name)
+            )
+            self._ridesharing_services_legacy.append(service)
+
+    def _init_class(self, cls, arguments):
+        """
+        Create an instance of a provider according to config
+        :param cls: provider class in Jormungandr found in config file
```

Comment:
Maybe document here which `cls` arguments are valid, otherwise it is hard to understand when the code is intended to work

Reply:
Ok.
I added tests  [here](https://github.com/CanalTP/navitia/pull/3295/commits/9a2556fef6b002e0ed320d66585cd0001ce1254f#diff-8c4a3e8fb4eba3a2374d7ce75213b10aaf174459bd873796d7a526a9ec073e22R161)